File: Colors.py
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        #Approach used is based on dutch algorithm
        #Space:O(1)
        #Time:O(n)
        left = 0
        curr = 0
        right = len(nums)-1
        while(curr<=right):
            if nums[curr] == 0:
                nums[curr],nums[left] = nums[left],nums[curr]
                curr+=1
                left+=1
            elif nums[curr] == 2:
                nums[curr],nums[right] = nums[right],nums[curr]
                # curr+=1
                right-=1
            else:
                curr+=1
        
        

File: colors.java
//time = o(n) space =o(1)
class Solution {
    public void swap(int[] nums, int a, int b) {
        int t = nums[a];
        nums[a] = nums[b];
        nums[b] = t;
    }

    public void sortColors(int[] nums) {
        int l = 0;
        int h = nums.length - 1;
        int m = 0;

        while (m <= h) {
            if (nums[m] == 0) {
                swap(nums, l, m);
                m++;
                l++;
            } else if (nums[m] == 2) {
                swap(nums, m, h);
                h--;
            } else {
                m++;
            }
        }
    }
}
